pkg/compact/blocks_cleaner.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// BlocksCleaner is a struct that deletes blocks from bucket
+// which are marked for deletion.
+type BlocksCleaner struct {
+	logger        log.Logger
+	deleteDelay   time.Duration
+	bkt           objstore.Bucket
+	blocksCleaned prometheus.Counter
+}
+
+// NewBlocksCleaner creates a new BlocksCleaner.
+func NewBlocksCleaner(logger log.Logger, bkt objstore.Bucket, deleteDelay time.Duration, blocksCleaned prometheus.Counter) *BlocksCleaner {
+	return &BlocksCleaner{
+		logger:        logger,
+		deleteDelay:   deleteDelay,
+		bkt:           bkt,
+		blocksCleaned: blocksCleaned,
+	}
+}
+
+// DeleteMarkedBlocks reads deletion-mark.json file in block to check when the block was marked for deletion and
+// deletes the block deleteDelay duration after block is marked for deletion.
+func (s *BlocksCleaner) DeleteMarkedBlocks(ctx context.Context) error {
+	return s.bkt.Iter(ctx, "", func(name string) error {
```

Comment:
This is really antipattern. Proper iteration is hard. That's why we build a metaFetcher framework to do it efficiently, cache on the way etc.... And now we start again. Also we already have all those detected by our deletion mark filter, right?

Reply:
You did not resolve my comment and just "resolved it". Please don't do that. Try to respond at least why you are ignoring. ):

Reply:
replied in this comment: https://github.com/thanos-io/thanos/pull/2136#issuecomment-593354806

Resolving this thread now.